packages/edit-site/src/index.js

```diff
@@ -25,43 +25,98 @@ import { store as editSiteStore } from './store';
 import { unlock } from './lock-unlock';
 import App from './components/app';
 
-const { registerCoreBlockBindingsSources } = unlock( editorPrivateApis );
+const { registerCoreBlockBindingsSources } = unlock(editorPrivateApis);
+
+/**
+ * Checks if the given path exists.
+ *
+ * @param {string} path The path to check.
+ * @return {boolean} True if the path exists, false otherwise.
+ */
+async function checkPathExists(path) {
+	const response = await fetch(`/wp-json/wp/v2/templates${path}`);
+	return response.ok;
+}
```

Comment:
Interesting idea, though I this can be implemented as part of the routing code to avoid the need to `fetch` asynchronously. In the site editor the routing is implemented here:
https://github.com/WordPress/gutenberg/blob/358fe48b742fd52fa0193663f4d7ae9f9a839aca/packages/edit-site/src/components/app/index.js#L53-L60

The `routes` come from this file:
https://github.com/WordPress/gutenberg/blob/trunk/packages/edit-site/src/components/site-editor-routes/index.js

The implementation is using the `router` package as a library. I can see that there's some handling for 404 here:
https://github.com/WordPress/gutenberg/blob/8e048a88af626169423f76136c264234819aef3b/packages/router/src/router.tsx#L163-L172

I'm not familiar with the code though, so I don't know how to make this display something. Perhaps it should be possible to define a `404` route in the site editor package